### DISCUSSION TITLE
Unbounded Memory Consumption During Collection Loading and Indexing

### DISCUSSION THREAD

**Discussion by ggtaro** [milvus-io/milvus#34639] [Q&A and General discussion]
Hi team. We are experiencing the following issue and any help would be greatly appreciated. Thanks.

## Description

We are experiencing severe memory consumption issues when loading a large collection and indexing new entities to the collection in Milvus standalone (launched via docker compose). The system consumes all available RAM at collection load time until the Linux OOM killer terminates the process.
We are using DISKANN index, and recently started experiencing a behavior where we are not able to load or index new entities into one of our bigger collections (~13 million entities). Every time we try to load the collection, milvus-standalone process eats all the memory (more than 300GB RAM) then process is terminated due to OOM. However, we have other collections that are a bit smaller (~10mm entities with similar fields and vectors, also DISKANN) and those mostly load fine (although also a bit heavy at loading and indexing time, but at least they have been loading).
In total we have ~50mm entities, 768dim vectors, other fields are ~0.5KB per entity, DISKANN index (default configs); spread over about 20 collections, bigger ones have close to 10mm entities. Biggest one is the one (above) with 13mm entities. Until recently, we were able to run ~40mm entities on a machine with three time less ram (120GB and 30 cores, DISKANN), but now even making the machine bigger is not helping. 

## Environment

- Milvus version: 2.4.0
- Deployment method: Docker Compose (Standalone)
- Host OS: Ubuntu 22.04
- RAM: 360GB
- CPU: 90 cores
- Storage: 1.4TB nvme

## Steps to Reproduce

1. Pre-existing collections (5 collections with 9mm entities in each; 768 vector dim, ~0.5kb metadata per entity; index to diskann)
2. Create the test collection (10mm entity index, 768 vector dim, ~0.5kb metadata per entity)
3. Build indices (DISKANN)
4. Try adding & indexing 3mm more entities
6. Try loading 13mm collection after milvus restart.

## Current Behavior

- During collection loading or index building, Milvus consumes all available system memory. Initially the OOM happened during indexing of new batch of data (small batches), at some point the process started consuming RAM, then machine went OOM and process died
- Next we tried to load the collection, and now it would not load and loading itself causes OOM and process dies.
- Before trying to load this 13mm sized collection machine has ~300GB RAM available, and after starting to load all of it gets consumed until process dies.
- We tried loading collections one by one, and all collections except this last one seem to load (although they too are resource heavy at load and index time, but tolerable and don't go full OOM)
- We tried setting hard memory limits in Docker Compose, but that results in the container being killed when the limit is reached, which is not a desirable solution.

## Questions:

1. Are there any known reasons why collection loading and indexing (minimal batches) with DISKANN can cause unbounded (or at least un-proportional to collection size) RAM usage?
2. Are there specific config settings that we can tune to force loading and indexing processes to consume less memory? (without sacrificing the query performance, we are OK sacrificing indexing/loading speeds for stability)
3. Given that our other ~10mm sized collections are able to load (also lots of RAM usage, but not all of it, and they load in the end), are there any steps we can take to check why this specific collection is behaving this way? Is there anything in DISKANN default configs that cause big change upon reaching a specific size?
4. Separately, when using DISKANN, does the non-vector scalar data (and all metadata) also get loaded into memory? How can we estimate the RAM consumption total? (from what I understand, DISKANN index will be ~1/5th of flat vector index, but I don't know what else gets loaded into RAM; should I add all scalar data to the calculation?) Can I offload non-index data to disk instead of RAM?
5. If collection gets large, should we split it into multiple collections? What threshold?

Any guidance on resolving these issues would be greatly appreciated. We're looking for a solution that allows us to work with large collections without requiring unreasonable amounts of RAM or risking system instability. Thanks.

**Comment by yhmo** [comment]
So, you are using a Milvus standalone.

We strongly recommend using Milvus cluster to process huge dataset. For standalone, all the work nodes(indexnode, querynode, datanode) are running in one process. All these work nodes might do heavy tasks at the same time, compactions, build index, load segments, etc. The server might hit a spike of memory and OOM.

So far as I know, there is no memory issue for DISKANN. I think the OOM you encountered is caused by standalone.

Scalar data is loaded into memory when you call collection.load(). If the scalar data is 0.5kb per entity, theoretically, memory usage of 10 million entities is 10M * 0.5KB = 5GB.

**Reply by ggtaro** [reply]
Hi,

Thank you for the response. We considered your answer and will be moving our collections to a milvus-cluster.
In that regard, I have the following questions
1. **If our milvus data is on Google Buckets, is it possible to move milvus to cluster setup and just point it to the existing an Google Bucket?** 
  1.1. I understand that all the metadata and schemas are in etcd and without etcd data we cannot simply point milvus to the existing bucket. **Is it possible to extract the etcd data (from a standalone or milvus cluster) in a way that would allow us to re-use our existing buckets?**
  1.2. I understand that there is a Milvus Backup Tool, but it seems like it also involves backing up the data from the buckets and reuploading them. **Does the Backup Tool allow for "etcd only" backup such that we skip the data backup?** I am assuming that having proper etcd data should be enough without the overhead of copying and re-uploading the bucket data.
  1.3. **Is it possible to do the same (backup etcd, but not data) for an existing running cluster and then give that etcd data to a different/new cluster and point that cluster to the existing bucket?**

2. We will be using GKE clusters. **Is there any recommended way for milvus to backup etcd data in GKE?** i.e. does your answer for first question change based on cluster being on GKE?

Our ideal scenario is having etcd backups readily available such that if we move the clusters around we can simply use the etcd backup and point the cluster to the existing data bucket. **The goal is to easily try out multiple cluster setups and settings until we optimize it for our performance and budget needs** including possibly having multiple milvus clusters running at the same time pointing to the same data bucket (for query only). Having the trial procedure be lightweight (only etcd backup dependent and not full data copying) would be very helpful.

Please let me know if any of the above is possible, and if not, what are the suggested workarounds.
Thank you for the responses.

**Reply by xiaofan-luan** [reply]
> Hi,
> 
> Thank you for the response. We considered your answer and will be moving our collections to a milvus-cluster. In that regard, I have the following questions
> 
> 1. **If our milvus data is on Google Buckets, is it possible to move milvus to cluster setup and just point it to the existing an Google Bucket?**
>    1.1. I understand that all the metadata and schemas are in etcd and without etcd data we cannot simply point milvus to the existing bucket. **Is it possible to extract the etcd data (from a standalone or milvus cluster) in a way that would allow us to re-use our existing buckets?**
>    1.2. I understand that there is a Milvus Backup Tool, but it seems like it also involves backing up the data from the buckets and reuploading them. **Does the Backup Tool allow for "etcd only" backup such that we skip the data backup?** I am assuming that having proper etcd data should be enough without the overhead of copying and re-uploading the bucket data.
>    1.3. **Is it possible to do the same (backup etcd, but not data) for an existing running cluster and then give that etcd data to a different/new cluster and point that cluster to the existing bucket?**
> 2. We will be using GKE clusters. **Is there any recommended way for milvus to backup etcd data in GKE?** i.e. does your answer for first question change based on cluster being on GKE?
> 
> Our ideal scenario is having etcd backups readily available such that if we move the clusters around we can simply use the etcd backup and point the cluster to the existing data bucket. **The goal is to easily try out multiple cluster setups and settings until we optimize it for our performance and budget needs** including possibly having multiple milvus clusters running at the same time pointing to the same data bucket (for query only). Having the trial procedure be lightweight (only etcd backup dependent and not full data copying) would be very helpful.
> 
> Please let me know if any of the above is possible, and if not, what are the suggested workarounds. Thank you for the responses.

You can not simply backup "etcd" or "S3", you have to backup both.
There are two way to make sure meta and data is consistent:

1. stop the cluster, physically backup etcd and then backup S3.
2. use backup tool to do that.

You can not directly copy S3 or etcd disk since this may cause consistency problem
